tests/unit/lms/views/exceptions_test.py

```diff
@@ -1,112 +1,138 @@
-    expected_result = {"message": "Page not found"}
+    @pytest.fixture
+    def exception(self):
+        return HAPIError("hapi_error_test_explanation")
 
 
-class TestForbidden(ExceptionViewTest):
-    view = exceptions.forbidden
-    exception = None
+class TestValidationError:
+    def test_it(self, assert_response, exception, exception_views):
+        assert_response(
+            exception_views.validation_error(),
+            status=exception.status_int,
+            error=exception,
+        )
 
-    response_status = 403
-    expected_result = {"message": "You're not authorized to view this page"}
+    @pytest.fixture
+    def exception(self):
+        return ValidationError(sentinel.messages)
 
 
-class TestHTTPClientError(ExceptionViewTest):
-    view = exceptions.http_client_error
-    exception = HTTPBadRequest("This is the error message")
+class TestReusedToolGUIDError:
+    def test_it(self, assert_response, exception_views, pyramid_request):
+        assert_response(exception_views.reused_tool_guid_error(), 200)
 
-    response_status = 400
-    expected_result = {"message": exception.args[0]}
+        pyramid_request.context.js_config.enable_error_dialog_mode.assert_called_with(
+            error_code=JSConfig.ErrorCode.REUSED_TOOL_GUID,
+            error_details={
+                "existing_tool_consumer_guid": sentinel.existing_guid,
+                "new_tool_consumer_guid": sentinel.new_guid,
+            },
+        )
 
+    @pytest.fixture
+    def exception(self):
+        return ReusedConsumerKey(sentinel.existing_guid, sentinel.new_guid)
 
-class TestHAPIError(ExceptionViewTest):
-    view = exceptions.hapi_error
-    exception = HAPIError("This is the error message")
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.context = create_autospec(
+            LTILaunchResource,
+            instance=True,
+            spec_set=True,
+            js_config=create_autospec(
+                JSConfig,
+                instance=True,
+                spec_set=True,
+                ErrorCode=JSConfig.ErrorCode,
+            ),
+        )
+        return pyramid_request
 
-    response_status = 500
-    expected_result = {"message": "This is the error message"}
 
+class TestError:
+    def test_it(self, assert_response, exception_views):
+        assert_response(
+            exception_views.error(),
+            500,
+            message="Sorry, but something went wrong. The issue has been "
+            "reported and we'll try to fix it.",
+        )
 
-class TestValidationError(ExceptionViewTest):
-    view = exceptions.validation_error
-    exception = ValidationError(mock.sentinel.messages)
+    @pytest.fixture
+    def exception(self):
+        return RuntimeError()
 
-    response_status = 422
-    expected_result = {"error": exception}
 
+@pytest.fixture
+def exception_views(exception, pyramid_request):
+    return ExceptionViews(exception, pyramid_request)
```

Comment:
> I find how this all hangs together a bit difficult to follow. Is this a good example of too much pytest where a simpler solution exists?
>
> ...
> 
> This would do a few nice things for us:
> 
> * Each method becomes decoupled from all the rest
> * It's a bit more _given, when, then_  / _arrange, act, assert_ style
> * How it works is local to the function
> * We could have a single test class `TestExceptionViews` with multiple `test_<method>` type methods which would shorten the file a lot
> 
> This leads to very short regularised tests like:
> 
> ```python
> class TestExceptionViews:
>     def test_notfound(self, assert_response, pyramid_request):
>         exception = httpexceptions.HTTPNotFound()
> 
>         response = ExceptionViews(exception, pyramid_request).notfound()
> 
>         assert_response(response, status=404, message="Page not found")
> 
>     def test_forbidden(self, assert_response, pyramid_request):
>         exception = httpexceptions.HTTPForbidden()
> 
>         response = ExceptionViews(exception, pyramid_request).forbidden()
> 
>         assert_response(
>             response, status=403, message="You're not authorized to view this page"
>         )
>     ...
> ```

Yeah, I think this is better.

The problem with the previous (subclass-based) solution was that it was a custom contraption that you had to understand in order to understand these tests, and it coupled all the tests together. The pytest fixtures-based solution is better in that it's less novel (it's using a technique straight out of the pytest docs that's used all over our tests) but it's still a contraption, just a pytest-based contraption, the tests are still more complicated than they need to be. It's also _less_ tightly coupled than the subclass-based approach but still coupled. I think simply inlining the fixtures is an improvement.

Done here: https://github.com/hypothesis/lms/pull/3157/commits/91df7dab6ead1a77dd9f2a087593fd051c1b4374